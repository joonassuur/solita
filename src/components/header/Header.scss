@import "../../index.scss";
@import "../../variables.scss";

#header-container {
  position: relative;
  overflow: hidden;
  .header {
    @extend .triangles;
    
    .container {
      display: flex;
      padding: 0 1em;
      justify-content: space-between;
    }
    display: flex;
    flex-direction: column;
    padding: 1em 0 0;
    background-color: $bgLight;
    justify-content: space-between;
    .description,
    .cart-display .cart {
      font-style: italic;
    }
    .logo {
      font-family: "Pacifico", sans-serif;
      font-size: 4em;
      line-height: 1em;
      color: $colorDark;
      text-decoration: none;
      cursor: pointer;
      margin-right: 0.5em;
    }
    .store-name,
    .cart-display {
      align-items: center;
      display: flex;
      .cart-icon {
        cursor: pointer;
        margin-left: 1em;
      }
      .cart-icon span {
        background: $bgDark;
        color: $colorLight;
        border-radius: 50px;
        padding: 8px;
      }
    }
  }
}

@media (max-width: $md) {
  .description {
    display: none;
  }
}
