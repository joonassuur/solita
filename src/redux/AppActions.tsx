import { createAction, createAsyncThunk } from "@reduxjs/toolkit";
import products from "../mockedAPI/products.json";

export const fetchStoreData = createAsyncThunk(
  "@app: fetchStoreData",
  async () => {
    // fake the api call
    const res = products;
    return res;
  }
);
export const modifyCart = createAction("@app: modifyCart", (item) => {
  return { payload: item };
});
export const toggleModal = createAction("@app: toggleModal", (isOpen) => {
  return { payload: isOpen };
});
