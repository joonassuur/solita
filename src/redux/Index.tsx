export {
  fetchStoreData,
  modifyCart,
  toggleModal,
} from "./AppActions";

export { getProducts, getCart, getIsModalOpen } from "./Selectors";
