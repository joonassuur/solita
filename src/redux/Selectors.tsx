import { StoreItem } from "../types/Types";
export const getProducts = (state: { app: { products: StoreItem[] } }) =>
  state.app.products;
export const getCart = (state: { app: { cart: StoreItem[] } }) =>
  state.app.cart;
