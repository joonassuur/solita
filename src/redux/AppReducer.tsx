import { createSlice } from "@reduxjs/toolkit";
import { fetchStoreData, modifyCart, toggleModal } from "../redux/AppActions";
import { Store } from "../types/Types";

import _ from "lodash";

const initialState: Store = {
  products: [],
  cart: [],
  isModalOpen: false,
};

const app = createSlice({
  name: "app",
  initialState,
  reducers: {},
  extraReducers: {
    [fetchStoreData.fulfilled.toString()]: (state, { payload }) => {
      state.products = payload.products;
    },
    [fetchStoreData.rejected.toString()]: (state) => {
      return state;
    },
    [modifyCart.toString()]: (state, { payload }) => {
      const clonedCart = _.cloneDeep(state.cart);
      const existingItemFilter = clonedCart.filter(
        (cartItem) => cartItem.id === payload.id
      );
      
      if (existingItemFilter.length > 0) {
        clonedCart.map((cartItem) =>
          cartItem.id === existingItemFilter[0].id
            ? (cartItem.quantity = payload.quantity)
            : false
        );
      } else {
        clonedCart.push(payload);
      }

      state.cart = clonedCart;
    },
    [toggleModal.toString()]: (state, { payload }) => {
      state.isModalOpen = payload;
    },
  },
});

export default app;
