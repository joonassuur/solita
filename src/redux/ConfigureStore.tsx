import { createStore, applyMiddleware  } from "redux";
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';

import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import rootReducer from "./RootReducer";

const persistConfig = {
  key: "root",
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export let store = createStore(persistedReducer, composeWithDevTools(applyMiddleware(thunk)));
export let persistor = persistStore(store);
