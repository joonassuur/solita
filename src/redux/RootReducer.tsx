import { combineReducers } from "redux";
import app from "./AppReducer";

const rootReducer = combineReducers({
  app: app.reducer,
});
export default rootReducer;
