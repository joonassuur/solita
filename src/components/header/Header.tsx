import React from "react";
import { useSelector } from "react-redux";
import { getCart } from "../../redux/Index";

import "./Header.scss";
interface PropsInterface {
  navigateToStore: (route?: string) => void;
  navigateToCart: (route?: string) => void;
}
const Header: React.FC<PropsInterface> = ({
  navigateToCart,
  navigateToStore,
}) => {
  const { cartTotalCost, cartQuantityString } = useSelector(getCart);

  return (
    <header id="header-container">
      <div className="header">
        <div className="container">
          <div className="store-name">
            <div className="logo" onClick={() => navigateToStore()}>
              whee
            </div>
            <div className="description">
              The most definitive shape store in the world
            </div>
          </div>
          <div className="cart-display">
            <div className="cart">
              <div className="cart-quantity">{cartQuantityString}</div>
              <div className="cart-cost">{`Total: ${cartTotalCost} â‚¬`}</div>
            </div>
            <div className="cart-icon" onClick={() => navigateToCart()}>
              <span className="material-icons">shopping_cart</span>
            </div>
          </div>
        </div>
      </div>
    </header>
  );
};

export default Header;
