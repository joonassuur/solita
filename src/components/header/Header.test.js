import React from "react";
import { shallow } from "enzyme";

import Header from "./Header";
import { getCart } from "../../redux/Selectors";

jest.mock("react-redux", () => ({
  useSelector: jest.fn((fn) => fn()),
}));
jest.mock("../../redux/Selectors");
getCart.mockReturnValue({ cartQuantityString: 2, cartTotalCost: 1798 });

describe("Header Component", () => {
  const props = {
    navigateToCart: jest.fn(),
    navigateToStore: jest.fn(),
  };
  const component = shallow(<Header {...props}/>);

  it("Should simulate cart icon click", () => {
    const navigateToCart = jest.fn(() => {});

    component.find(".cart-icon").simulate("click", navigateToCart());
    expect(navigateToCart).toBeCalled();
  });

  it("Should display .cart-quantity element text output as '2'", () => {
    expect(component.find(".cart-quantity").text()).toEqual("2");
  });

  it("Should display .cart-cost element text output as 'Total: 1798 €'", () => {
    expect(component.find(".cart-cost").text()).toEqual("Total: 1798 €");
  });

  it("Should match Header component snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
