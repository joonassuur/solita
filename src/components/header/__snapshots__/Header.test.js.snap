// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Should match Header component snapshot 1`] = `ShallowWrapper {}`;
